extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusher.java

```diff
@@ -124,8 +124,18 @@ public DataSegment pushToPath(File inDir, DataSegment segment, String storageDir
         fullyQualifiedStorageDirectory.get(),
         storageDirSuffix
     );
-
-    final String storageDir = StringUtils.format("%s/%s", fullyQualifiedStorageDirectory.get(), storageDirSuffix);
+    
+    final String outIndexFilePath;
+    if (storageDirSuffix.endsWith("index.zip")) {
```

Comment:
Could you please add some comments for both the if and else here? I am not sure why the `storageDirSuffix` would contain `index.zip` in it.

Reply:
After taking a look at the linked issue #14972 , it seems that the method `pushToPath` is always being called with the directory name except in `DeepStorageIntermediaryDataManager` where it is called with the actual filename.

So the problem is not specific to `HdfsDataSegmentPusher` and might occur for other deep stores as well.

The right thing to do is either to fix up `DeepStorageIntermediaryDataManager` or add a new `push` method which takes the filename rather than directory.

Reply:
@kfaraz Thanks for looking into this! 

My first intuition was also trying to fix `DeepStorageIntermediaryDataManager` by passing a file path contains `index.zip` suffix here
https://github.com/apache/druid/blob/3ae5e978012e7a553e6ad3007eb210ab8da5b6fc/indexing-service/src/main/java/org/apache/druid/indexing/worker/shuffle/DeepStorageIntermediaryDataManager.java#L78

But after checking other implementations of `DataSegmentPusher`, I found they all call method [pushToPath](https://github.com/apache/druid/blob/3ae5e978012e7a553e6ad3007eb210ab8da5b6fc/processing/src/main/java/org/apache/druid/segment/loading/DataSegmentPusher.java#L71) in their [push](https://github.com/apache/druid/blob/3ae5e978012e7a553e6ad3007eb210ab8da5b6fc/processing/src/main/java/org/apache/druid/segment/loading/DataSegmentPusher.java#L69) method. Some of them format segment file path contains `index.zip` suffix and pass it by parameter `storageDirSuffix` to method `pushToPath`, see below
1. `AzureDataSegmentPusher`
https://github.com/apache/druid/blob/3ae5e978012e7a553e6ad3007eb210ab8da5b6fc/extensions-core/azure-extensions/src/main/java/org/apache/druid/storage/azure/AzureDataSegmentPusher.java#L124-L125
2. `HdfsDataSegmentPusher`
https://github.com/apache/druid/blob/279b3818f0e02f6a2ff389532f1e6118598a7c24/extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusher.java#L108-L115

Other implementations format `storageDirSuffix` by calling [getStorageDir](https://github.com/apache/druid/blob/279b3818f0e02f6a2ff389532f1e6118598a7c24/processing/src/main/java/org/apache/druid/segment/loading/DataSegmentPusher.java#L89) in their `push` methods.

To fix the issue, I'd prefer to fix the inconsistent usages of `pushToPath` in method `AzureDataSegmentPusher#push` and `HdfsDataSegmentPusher#push`. We should always pass a dir to parameter `storageDirSuffix` of method `pushToPath` and format segment file path with suffix `index.zip` in method `pushToPath`. So, we don't need to change `DeepStorageIntermediaryDataManager` as it passes a dir to method `pushToPath`. Does it make sense to you?

Reply:
@kfaraz Friendly bump!

Reply:
The function name is a bit confusing since storageDirSuffix is really a suffix to the filename. I looked at various implementations and this is what you can do to simplify the logic. Rather than `push` call `pushToPath` directly, both of these functions can call a separate third private function. `push` will create the right suffix by adding index.zip and so would `pushToPath`. `pushToPath` will add the suffix in the same way as org.apache.druid.storage.s3.S3Utils#constructSegmentPath

How does that sound?

Reply:
@abhishekagarwal87 Thanks for your comment!

> The function name is a bit confusing since storageDirSuffix is really a suffix to the filename.

I think `storageDirSuffix` shouldn't be the suffix to the filename but instead as the prefix dir of file `index.zip`.

I have concern that adding a third method will need to touch all implementations. I think it's ok to reuse `pushToPath` method inside `push` method as long as it complies with passing a storageDirSuffix not file path to method `pushToPath`.

Currently, only two implementations, `HdfsDataSegmentPusher` and `AzureDataSegmentPusher`, violate to pass file path  to their method `pushToPath` methods. Other implementations only format `storageDirSuffix` and reuse method `pushToPath` to format file path with `index.zip` file.

We can also enhance the method doc of `pushToPath`
https://github.com/apache/druid/blob/3ae5e978012e7a553e6ad3007eb210ab8da5b6fc/processing/src/main/java/org/apache/druid/segment/loading/DataSegmentPusher.java#L70-L75

Reply:
Oh, I was suggesting refactoring just the HDFSDataSegmentPusher so you don't need to do that "endsWith" check in this class.

Reply:
@abhishekagarwal87 Oh I see. Updated.